pkg/webhook/policy.go

```diff
@@ -313,8 +313,32 @@ func (h *validationHandler) validateConstraint(ctx context.Context, req admissio
 	return false, nil
 }
 
-// traceSwitch returns true if a request should be traced
 func (h *validationHandler) reviewRequest(ctx context.Context, req admission.Request) (*rtypes.Responses, error) {
+	return h.withTracing(ctx, req, func(tracing opa.QueryOpt) (*rtypes.Responses, error) {
```

Comment:
I like the idea of separation of concerns.

I could see the set of conditional execution flags being extended in the future. In that scenario, would it be better to have a separate function such as `GetReviewArgs()` that dynamically generates the arguments for review?

That would avoid multiple layers of wrapping.

Another way I could see doing this would be to have each argument defined explicitly and then passed in, e.g. something like:

```go
withTracing := makeTracingArg()
otherThing := makeOtherArg()
... other args
```

This second method has the advantage of being more explicit about what arguments are being set, but the disadvantage of being a bit more expensive to maintain.

Reply:
in general I like wrapping because it takes care of the set/unset flow, multiple things could be handled by either adding options:

```
withA a {
  withB b {
    someCall(a + b)
  }
}
```

or with passing in the option:
```
options = {}
withA(options) {
  withB(options) {
     someCall(options)
  }
}
```
which I kinda like more since it removes the option adding inside

Reply:
here it looks like the interface is `...QueryOpt` so if we add more calls it would be
`someCall(a, b)` which is pretty clean compared to having a QueryOpt list and then passing that along and adding to it

Reply:
I'm unclear what the syntax is for the suggestion:

```
withA a {
  withB b {
    someCall(a + b)
  }
}
```

I'm guessing `a` and `b` are integers and it's a bitmask-style argument set, which is why we can add options?

The two biggest reasons I have for being leery of nested function calls are:

1) The extra nesting on the call stack will make tracebacks harder to read as each extra argument will add a line to the traceback
2) As the number of arguments increases, the nesting depth increases, making the code harder to grok.
 
The ability to set/unset is cool, but not currently necessary for this code.

Re: functional args: have you read [Dave Cheny's post](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis)? It's a good read.

Reply:
Each of the "with" methods could append to the list internally, which would look like:

```
var args []QueryOpt
WithA(args)
WithB(args)
query(args...)
```